registry/backoff.go

```diff
@@ -0,0 +1,93 @@
+package registry
+
+import (
+	"errors"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+)
+
+const (
+	initialBackoff = 500 * time.Millisecond
+	maxBackoff     = 10 * time.Second
+)
+
+var (
+	ErrTimeout = errors.New("http request timeout")
+)
+
+type backoffRoundTripper struct {
+	roundTripper               http.RoundTripper
+	initialBackoff, maxBackoff time.Duration
+	clock                      clockwork.Clock
+}
+
+// BackoffRoundTripper is a http.RoundTripper which adds a backoff for
+// throttling to requests. To add a total request timeout, use Request.WithContext.
+//
+// r              -- upstream roundtripper
+// initialBackoff -- initial length to backoff to when request request fails
+// maxBackoff     -- maximum length to backoff to between request attempts
+func BackoffRoundTripper(r http.RoundTripper, initialBackoff, maxBackoff time.Duration, clock clockwork.Clock) http.RoundTripper {
+	return &backoffRoundTripper{
+		roundTripper:   r,
+		initialBackoff: initialBackoff,
+		maxBackoff:     maxBackoff,
+		clock:          clock,
+	}
+}
+
+func (c *backoffRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	b := &backoff{
+		initial: c.initialBackoff,
+		max:     c.maxBackoff,
+	}
+	for {
+		// Try the request
+		resp, err := c.roundTripper.RoundTrip(r)
+		switch {
+		case err != nil && strings.Contains(err.Error(), "Too Many Requests (HAP429)."):
+			// Catch the terrible dockerregistry error here. Eugh. :(
+			fallthrough
+		case resp != nil && resp.StatusCode == http.StatusTooManyRequests:
+			// Request rate-limited, backoff and retry.
+			b.Failure()
+			// Wait until the next time we are allowed to make a request
+			c.clock.Sleep(b.Wait())
```

Comment:
Since i don't think there's any need to lock the backoff, I don't think we need a fancy sleep?